engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java

```diff
@@ -659,22 +657,17 @@ private BeanPathMetaConstraintProcessedUnit(Object bean, Path path, MetaConstrai
 
 		@Override
 		public boolean equals(Object o) {
-			if ( this == o ) {
-				return true;
-			}
-			if ( o == null || o.getClass() != BeanPathMetaConstraintProcessedUnit.class ) {
-				return false;
-			}
-
+			// no need to check for instance equality or if the calss match because of
+			// how this class is used in the set.
 			BeanPathMetaConstraintProcessedUnit that = (BeanPathMetaConstraintProcessedUnit) o;
 
 			if ( bean != that.bean ) {  // instance equality
```

Comment:
Can't you keep that one at the top? Let's avoid the cast if we can.

Reply:
there's `that` hiding in the if statement so cannot avoid the cast